MoreLinq/UnFold.cs

```diff
@@ -0,0 +1,87 @@
+// limitations under the License.
+#endregion
+
+namespace MoreLinq
+{
+    using System;
+    using System.Collections.Generic;
+
+    static partial class MoreEnumerable
+    {
+        /// <summary>
+        /// Returns a sequence generated by applying a state to the generator function, 
+        /// and from its result, determines if the sequence should have a next element and its value,
+        /// and the next state in the recursive call.
+        /// </summary>
+        /// <typeparam name="TState">Type of state elements.</typeparam>
+        /// <typeparam name="T">Type of the elements generated by the generator function.</typeparam>
+        /// <typeparam name="TResult">The type of the elements of the result sequence.</typeparam>
+        /// <param name="state">The initial state.</param>
+        /// <param name="generator">
+        /// Function that takes a state and produces an object containing all 
+        /// necessary data for the other functions (predicate, stateSelector, resultSelector).
+        /// </param>
+        /// <param name="predicate">
+        /// Function to determine if the unfolding should continue based on the current input.
+        /// </param>
+        /// <param name="stateSelector">
+        /// Function to select the state from the current input.
+        /// </param>
+        /// <param name="resultSelector">
+        /// Function to select the result from the current input.
+        /// </param>
+        /// <returns>A sequence containing the results generated by the resultSelector function.</returns>
+        /// <remarks>
+        /// This operator uses deferred execution and streams its results.
+        /// </remarks>
+
+        public static IEnumerable<TResult> Unfold<TState, T, TResult>(
+            TState state,
+            Func<TState, T> generator,
+            Func<T, bool> predicate,
+            Func<T, TState> stateSelector,
+            Func<T, TResult> resultSelector)
+        {
+            if (generator == null) throw new ArgumentNullException(nameof(generator));
+            if (predicate == null) throw new ArgumentNullException(nameof(predicate));
+            if (stateSelector == null) throw new ArgumentNullException(nameof(stateSelector));
+            if (resultSelector == null) throw new ArgumentNullException(nameof(resultSelector));
+
+            return UnfoldImpl(state, generator, predicate, stateSelector, resultSelector);
+        }
+
+
+        private static IEnumerable<TResult> UnfoldImpl<TState, T, TResult>(
+            TState state,
+            Func<TState, T> generator,
+            Func<T, bool> predicate,
+            Func<T, TState> stateSelector,
+            Func<T, TResult> resultSelector)
+        {
+            while (true)
+            {
+                var step = generator(state);
+
+                if (!predicate(step))
+                    yield break;
+
+                yield return resultSelector(step);
+                state = stateSelector(step);
```

Comment:
How do you get back the result to yield for the current iteration from the state for the next that is `step`? Suppose your seed is 0 and `generator` is a simple increment. Then `step` will be 1 for the _next_ iteration. Your `predicate` and `resultSelector` would have to account for that. For example, `resultSelector` will have to decrement to get the result, so at best, it'll feel clumsy. The `squaresLowerThan100` example from #246 will now read like this: 

```c#
var squaresLowerThan100 =
    Unfold(1, x => x + 1, x => ((x - 1) * (x - 1)) < 100, x => (x - 1) * (x - 1));
```

> where one function is an identity

Which function?